### PR TITLE
Feat: Added support for PostgreSQL schema in @voltagent/postgres

### PR DESCRIPTION
**Pull request by hyperion912**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://voltagent.dev/docs/community/contributing/#commit-convention

## Bugs / Features

- [x] Related issue(s) linked
- [x] Tests for the changes have been added
- [ ] Docs have been added / updated
- [x] Changesets have been added https://voltagent.dev/docs/community/contributing/#creating-a-changeset

fixes (issue) #763 

## What is the current behavior?
Currently, the PostgreSQL Memory Adapter (`@voltagent/postgres`) has the following limitations:

### 1. Fixed Schema (Breaking Limitation)
- **All tables are created in the `public` schema only**
- No way to specify a custom PostgreSQL schema
- Cannot organize tables by application or environment
- Conflicts with multi-tenant architectures
- Cannot follow PostgreSQL best practices for schema organization

### 2. No Schema Management
- No programmatic way to inspect database schema
- Cannot query table information and statistics
- Cannot manage tables within a schema
- No database maintenance operations

This creates friction for users who want to:
- **Multi-tenancy**: Separate data by schema per tenant
- **Organization**: Group Voltagent tables in a dedicated schema (e.g., `voltagent`, `ai_agent`, `chatbot`)
- **Security**: Use PostgreSQL schema-level permissions
- **Environments**: Use different schemas for dev/staging/prod in same database
- **Monitoring**: Monitor database health and table growth
- **Maintenance**: Perform optimization tasks programmatically

## What is the new behavior?

This PR adds **two major features**:

### Feature 1: Custom Schema Configuration

Users can now specify a **custom PostgreSQL schema** for all Voltagent tables:

```typescript
import { PostgreSQLMemoryAdapter } from '@voltagent/postgres';

// NEW: Specify custom schema
const adapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  schema: 'my_app_ai',           // Custom schema
  tablePrefix: 'voltagent'
});

// Tables will be created as:
// - my_app_ai.voltagent_users
// - my_app_ai.voltagent_conversations
// - my_app_ai.voltagent_messages
// - my_app_ai.voltagent_workflow_states

// BACKWARD COMPATIBLE: Default is 'public'
const defaultAdapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  // schema not specified = 'public' (existing behavior)
});
```

**Benefits:**
- **Multi-tenancy**: `tenant_a`, `tenant_b`, `tenant_c` schemas
- **Organization**: `voltagent`, `ai_agents`, `chatbot` schemas
- **Security**: Schema-level PostgreSQL permissions
- **Environments**: `dev`, `staging`, `prod` schemas in same DB
- **Isolation**: Keep Voltagent tables separate from application tables

**How it works:**
- Schema is automatically created if it doesn't exist
- All tables, indexes, and queries use the specified schema
- Proper SQL quoting for schema names (handles special characters)
- Falls back to `public` schema if not specified (backward compatible)

### Feature 2: Schema Management Functions

**12 new methods** for full programmatic control over the database schema:

#### Information Methods
1. **`getSchemaName(): string`** - Returns the current PostgreSQL schema being used
2. **`getTablePrefix(): string`** - Returns the table prefix for all managed tables
3. **`listTables(): Promise<string[]>`** - Lists all tables in the current schema
4. **`listManagedTables(): Promise<string[]>`** - Lists only tables managed by this adapter (with configured prefix)
5. **`tableExists(tableName: string): Promise<boolean>`** - Checks if a specific table exists

#### Inspection Methods
6. **`getTableStats(tableName: string)`** - Returns comprehensive table statistics:
   ```typescript
   {
     rowCount: number;        // Total rows in table
     totalSize: string;       // Total size (table + indexes)
     indexSize: string;       // Size of all indexes
     tableSize: string;       // Size of table data only
   }
   ```

7. **`getTableIndexes(tableName: string)`** - Returns all indexes with metadata:
   ```typescript
   {
     indexName: string;       // Name of the index
     columnNames: string[];   // Columns included in index
     isUnique: boolean;       // Whether index enforces uniqueness
     isPrimary: boolean;      // Whether this is the primary key
   }
   ```

8. **`getTableSchema(tableName: string)`** - Returns column information:
   ```typescript
   {
     columnName: string;               // Column name
     dataType: string;                 // PostgreSQL data type
     isNullable: boolean;              // Can be NULL
     defaultValue: string | null;      // Default value
     characterMaximumLength: number | null;  // Max length (for strings)
   }
   ```

#### Management Methods
9. **`dropManagedTables(): Promise<void>`** - Drops all managed tables (WARNING: destructive)
10. **`truncateManagedTables(): Promise<void>`** - Truncates all managed tables (deletes data, keeps structure)
11. **`executeRawQuery<T>(sql: string, params?: any[]): Promise<T[]>`** - Executes custom SQL queries
12. **`vacuumAnalyzeManagedTables(): Promise<void>`** - Optimizes all managed tables for better performance

### Example Usage

```typescript
import { PostgreSQLMemoryAdapter } from '@voltagent/postgres';

// Example 1: Custom schema for organization
const adapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  schema: 'my_app_ai',      // Custom schema
  tablePrefix: 'voltagent'
});

// Example 2: Multi-tenant setup
const tenantAdapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  schema: `tenant_${tenantId}`,  // Dynamic schema per tenant
  tablePrefix: 'chat'
});

// Example 3: Environment-based schemas
const adapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  schema: process.env.NODE_ENV,  // 'dev', 'staging', 'prod'
  tablePrefix: 'voltagent'
});

// Get schema information (NEW)
const schema = adapter.getSchemaName(); // 'my_app_ai'
const prefix = adapter.getTablePrefix(); // 'voltagent'

// List all managed tables
const tables = await adapter.listManagedTables();
// ['voltagent_users', 'voltagent_conversations', 'voltagent_messages', 'voltagent_workflow_states']

// Check if a table exists
const exists = await adapter.tableExists('voltagent_conversations'); // true

// Get table statistics
const stats = await adapter.getTableStats('voltagent_messages');
console.log(`Table has ${stats.rowCount} rows and uses ${stats.totalSize}`);

// Get table indexes
const indexes = await adapter.getTableIndexes('voltagent_conversations');
indexes.forEach(idx => {
  console.log(`${idx.indexName}: ${idx.columnNames.join(', ')} ${idx.isPrimary ? '(PRIMARY)' : ''}`);
});

// Get table schema
const schema = await adapter.getTableSchema('voltagent_conversations');
schema.forEach(col => {
  console.log(`${col.columnName}: ${col.dataType} ${col.isNullable ? 'NULL' : 'NOT NULL'}`);
});

// Maintenance operations
await adapter.vacuumAnalyzeManagedTables(); // Optimize all tables

// Custom queries (advanced)
const results = await adapter.executeRawQuery<{count: number}>(
  'SELECT COUNT(*) as count FROM voltagent_messages WHERE created_at > $1',
  [new Date('2024-01-01')]
);

// Cleanup (WARNING: destructive - use with caution)
await adapter.truncateManagedTables(); // Delete all data
await adapter.dropManagedTables();     // Drop all tables
```

### Source Code Changes

**File: `packages/postgres/src/memory-adapter.ts`**

#### Schema Configuration Support
- Added `schema` parameter to `PostgreSQLMemoryOptions` interface (line 57)
- Added `private schema: string` field to class (line 74)
- Updated `constructor` to accept and store schema configuration (line 81)
- Modified `getTableName()` method to properly quote and prefix tables with schema (lines 119-124)
- Updated `initialize()` method to create custom schema if specified (lines 142-145)
- All SQL queries now use schema-aware table names via `getTableName()`

#### Schema Management Functions  
- Added 12 new public methods for schema management (lines 1167-1435)
- `getSchemaName()` and `getTablePrefix()` - Return current configuration
- Methods for listing tables, checking existence, getting stats/indexes/schema
- Maintenance methods: truncate, drop, vacuum, raw query execution
- All methods respect the configured `schema` and `tablePrefix`
- Methods use existing connection pool (no new connections)
- Proper error handling with transactions where appropriate
- TypeScript types for all return values

### Test Changes

**File: `packages/postgres/src/memory-adapter.spec.ts`**

- Added comprehensive test suite: "PostgreSQLMemoryAdapter - Schema Validation" (lines 900-1432)
  - Default schema behavior (public)
  - Custom schema with proper SQL quoting (`"schema_name"."table_name"`)
  - Schema creation SQL verification
  - Table prefix configuration
  - Table structure validation (columns, types, constraints)
  - Index creation verification
  - Connection string vs object configuration
  
- Added test suite: "PostgreSQLMemoryAdapter - Schema Management Functions" (lines 1435-1735)
  - All 12 new schema management functions
  - SQL query validation for each method
  - Error handling and edge cases

- **24 new unit tests** total
- Fixed mock initialization to properly handle migrations
- All tests use mocking (no real database required)
- Tests validate both return values and SQL query generation

### For Users

**Schema Configuration:**
- **Multi-tenancy support** - Isolate data per tenant using PostgreSQL schemas
- **Better organization** - Group Voltagent tables in dedicated schemas
- **Enhanced security** - Leverage PostgreSQL schema-level permissions
- **Environment separation** - Use different schemas for dev/staging/prod
- **Clean separation** - Keep Voltagent tables separate from application data
- **Backward compatible** - Defaults to `public` schema (no breaking changes)

**Schema Management:**
- **Full schema visibility** - Inspect tables, columns, indexes programmatically
- **Database monitoring** - Track table sizes, row counts, and growth
- **Maintenance automation** - VACUUM, cleanup, and optimization tasks
- **Admin interfaces** - Build custom dashboards and management tools
- **Debugging** - Quickly verify schema structure and data integrity

### For Developers
- **Better testability** - Comprehensive test coverage with clear examples
- **Type safety** - Full TypeScript support for all new methods
- **Consistent API** - Methods follow existing adapter patterns

## Testing

All tests pass successfully:

```bash
npm run test

✓ PostgreSQLMemoryAdapter - Core Functionality (17 tests)
✓ PostgreSQLMemoryAdapter - Schema Validation (8 tests)  
✓ PostgreSQLMemoryAdapter - Schema Management Functions (13 tests)

Tests: 38 passed (38 total)
```

### Test Coverage

**Schema Configuration:**
- Default schema handling (`public`) - backward compatibility
- Custom schema with proper SQL quoting (`"my_schema"."table"`)
- Schema creation verification (`CREATE SCHEMA IF NOT EXISTS`)
- Schema prefix in all table operations
- Public schema special handling (no prefix needed)

**Schema Management:**
- Table prefix configuration
- Table and index creation
- All 12 new schema management methods
- Error handling and edge cases
- SQL query validation

## Breaking Changes

**None.** This PR is **100% backward compatible**:

### Schema Configuration (Backward Compatible)
- `schema` parameter is **optional** in constructor
- Defaults to `'public'` (existing behavior) if not specified
- Existing adapters without `schema` parameter continue to work
- All tables continue to be created in `public` schema by default
- No changes required to existing code

### Schema Management Functions (Additive Only)
- All existing methods work unchanged
- New methods are additive only
- Existing code continues to work without modification

**Example - Existing code still works:**
```typescript
// This code works exactly as before (uses public schema)
const adapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  tablePrefix: 'voltagent'  // schema defaults to 'public'
});
```

## Migration Guide

No migration needed! Simply update to the latest version:

```bash
npm install @voltagent/postgres@latest
```

**To use custom schemas (optional):**
```typescript
const adapter = new PostgreSQLMemoryAdapter({
  connection: process.env.DATABASE_URL,
  schema: 'your_custom_schema',  // Add this line
  tablePrefix: 'voltagent'
});
```

New methods are available immediately on existing adapter instances.

### Key Points

1. **Custom Schema is Optional**: 
   - Defaults to `public` schema (existing behavior)
   - Users can opt-in to custom schemas when needed
   - 100% backward compatible

2. **Schema Isolation Benefits**:
   - Perfect for multi-tenant applications
   - Enables environment-based organization (dev/staging/prod)
   - Supports schema-level PostgreSQL permissions
   - Keeps Voltagent tables organized and separate

3. **Schema Management is Opt-in**: 
   - Users don't need to use management methods if they don't want to
   - Core functionality (conversations, messages, workflows) remains unchanged

4. **Security Considerations**: 
   - Custom schema requires appropriate PostgreSQL CREATE SCHEMA permission
   - `dropManagedTables()` and `truncateManagedTables()` are destructive but clearly documented with WARNING labels
   - `executeRawQuery()` allows custom SQL but respects the connection pool and schema configuration
   - All methods require the user to have appropriate PostgreSQL permissions

5. **Performance**: 
   - Schema configuration has zero performance impact
   - All methods reuse the existing connection pool
   - No additional connections or overhead
   - `vacuumAnalyzeManagedTables()` runs outside transactions (PostgreSQL requirement)

### CODE REVIEW

File: packages/postgres/src/memory-adapter.ts [hunk 1]
```diff
@@ -537,8 +561,8 @@ export class PostgreSQLMemoryAdapter implements StorageAdapter {
     try {
       await client.query("BEGIN");
 
-      const messagesTable = `${this.tablePrefix}_messages`;
-      const conversationsTable = `${this.tablePrefix}_conversations`;
+      const messagesTable = this.getTableName(`${this.tablePrefix}_messages`);
+      const conversationsTable = this.getTableName(`${this.tablePrefix}_conversations`);
```
Comment (Diluka) [inline packages/postgres/src/memory-adapter.ts:568]: https://github.com/hyperion912/voltagent/blob/564f9a9e24b1292f44955196011eeab3d16b4094/packages/postgres/src/memory-adapter.ts#L209-L213

Is this place where the name including the schema and a dot is appended to the index name?

Comment (hyperion912) [reply packages/postgres/src/memory-adapter.ts:568]: yes

Comment (Diluka) [reply packages/postgres/src/memory-adapter.ts:568]: Is this a good idea?

Comment (Diluka) [reply packages/postgres/src/memory-adapter.ts:568]: However, the following statement doesn't use double quotes to enclose the index name, so will the presence of a dot cause an error? Also, the previous index names didn't include a schema; will this cause a breaking change for users using the public default value?

Comment (hyperion912) [reply packages/postgres/src/memory-adapter.ts:568]: Yes, you are right. This will cause an error because `idx_"schemaName"`  will be treated as schema-qualified identifier.
Since schema name is not necessary in the index name, i will be using base tableName without schema for naming the indexes.
Also if the user is using public default value then nothing will be added to the tableName.

```  
private getTableName(tableName: string): string {
  if (this.schema && this.schema !== "public") {
        return `"${this.schema}"."${tableName}"`;
  }
  
  return tableName;
}
  ```
  So, this will not cause any breaking changes.
  
  Thank you for pointing it out.

**Comment by Diluka** [comment]
Have you considered keeping the default value of `schema` in this version as `undefined`? This would ensure 100% compatibility for existing users, with no changes whatsoever. This is especially important for databases where the default schema is not set to `public`.

**Comment by hyperion912** [comment]
> Have you considered keeping the default value of `schema` in this version as `undefined`? This would ensure 100% compatibility for existing users, with no changes whatsoever. This is especially important for databases where the default schema is not set to `public`.

I haven't considered the possibility for databases where default schema is not set `public`. Thank you for pointing it out. I will change the default value as undefined.

File: packages/postgres/src/memory-adapter.ts [hunk 2]
```diff
@@ -50,6 +50,12 @@ export interface PostgreSQLMemoryOptions {
    */
   tablePrefix?: string;
 
+  /**
+   * PostgreSQL schema to use for all tables
+   * @default "public"
```
Comment (Diluka) [inline packages/postgres/src/memory-adapter.ts]:
```suggestion
   * @default undefined
```

File: packages/postgres/src/memory-adapter.ts [hunk 3]
```diff
@@ -1139,13 +1166,352 @@ export class PostgreSQLMemoryAdapter implements StorageAdapter {
...
+     FROM information_schema.tables
+     WHERE ${schemaCondition}
+     AND table_type = 'BASE TABLE'
+     AND table_name LIKE $${this.schema ? 2 : 1}
+     ORDER BY table_name`,
+      params,
+    );
+
+    return result.rows.map((row) => row.table_name);
+  }
```
Comment (Diluka) [inline packages/postgres/src/memory-adapter.ts]: This is problematic. If the table prefix is ​​the same as other business tables, it will result in redundant tables. However, the tables in the adapter itself are fixed and do not require additional queries.

File: packages/postgres/src/memory-adapter.ts [hunk 4]
```diff
@@ -1139,13 +1166,352 @@ export class PostgreSQLMemoryAdapter implements StorageAdapter {
...
+      defaultValue: row.column_default,
+      characterMaximumLength: row.character_maximum_length,
+    }));
+  }
+
+  /**
+   * Drop all managed tables (with the configured table prefix)
+   * WARNING: This will delete all data in these tables
+   */
+  async dropManagedTables(): Promise<void> {
```
Comment (Diluka) [inline packages/postgres/src/memory-adapter.ts]: Similar to the one above, it's possible to drop other tables here.

**Review by Diluka — COMMENTED** [review]
1. The basic adapter section seems fine to me.

2. The newly added management section appears problematic; is it even necessary?  
  - @hyperion912  Why is this needed? This doesn't seem to be the responsibility of the MemoryAdapter.
  - @omeraplak What are your thoughts on this?

3. I'm not familiar with the final database structure migration section; I suggest separating it into a separate file in future versions. @omeraplak

File: .changeset/social-onions-play.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+"@voltagent/postgres": major
```
Comment (omeraplak) [inline .changeset/social-onions-play.md]:
```suggestion
"@voltagent/postgres": minor
```
Hey,
Did we introduce any breaking changes in this PR?
If not, we can go with a patch or minor version instead of a major one.

File: packages/postgres/src/memory-adapter.ts (hunk 4 above)
Comment (omeraplak) [reply packages/postgres/src/memory-adapter.ts]: Yeah, I agree. Do we actually need this method?

File: packages/postgres/src/memory-adapter.ts (hunk 3 above)
Comment (omeraplak) [reply packages/postgres/src/memory-adapter.ts]: Yeah, I agree. Do we actually need this method?

Comment (hyperion912) [reply packages/postgres/src/memory-adapter.ts]: You're right. I added those methods because the original feature request mentioned:
`Functions for managing tables within a specific schema`, so you could do things like list tables, drop tables, or migrate only within that schema.
I understood that as needing utility functions for managing tables within a schema, mainly for:
- Cleaning up test schemas during development
- Multi-tenant setups where each tenant has its own schema
- Running migrations or resets programmatically

But @Diluka , you’re absolutely correct that:
- Using LIKE for pattern matching could unintentionally include unrelated tables
- Table management belongs more to database administration than adapter logic
- The adapter already knows its table names, so dynamic discovery isn’t really necessary

It makes sense to remove these methods and keep the adapter focused on its core responsibilities.
Should I remove the functions entirely or move them into a new file?

File: .changeset/social-onions-play.md (hunk 5 above)
Comment (hyperion912) [reply .changeset/social-onions-play.md]: No, we didn't introduce any breaking changes.

File: packages/postgres/src/memory-adapter.ts (hunk 3 above)
Comment (Diluka) [reply packages/postgres/src/memory-adapter.ts]: I just read the issue, and copilot‘s rewriting of my original simple meaning is indeed a little ambiguous.

I suggest removing it. The way of data maintenance will vary according to different projects. There is no clear real scenario for this code. Except for integrated testing, I don‘t think anyone dares to use it. After all, data is priceless.

Comment (hyperion912) [reply packages/postgres/src/memory-adapter.ts]: Got it!

**Comment by omeraplak** [comment]
Thnak you so much @hyperion912 ⚡️
